cpp/src/arrow/stl.h

```diff
@@ -50,19 +52,33 @@ using BareTupleElement = typename std::remove_const<typename std::remove_referen
 template <typename T, typename Enable = void>
 struct ConversionTraits {};
 
-#define ARROW_STL_CONVERSION(c_type, ArrowType_)                                    \
-  template <>                                                                       \
-  struct ConversionTraits<c_type> : public CTypeTraits<c_type> {                    \
-    static Status AppendRow(typename TypeTraits<ArrowType_>::BuilderType& builder,  \
-                            c_type cell) {                                          \
-      return builder.Append(cell);                                                  \
-    }                                                                               \
-    static c_type GetEntry(const typename TypeTraits<ArrowType_>::ArrayType& array, \
-                           size_t j) {                                              \
-      return array.Value(j);                                                        \
-    }                                                                               \
-    constexpr static bool nullable = false;                                         \
-  };
+#define ARROW_STL_CONVERSION(CType_, ArrowType_)                                         \
+  template <>                                                                            \
+  struct ConversionTraits<CType_> : public CTypeTraits<CType_> {                         \
+    static Status AppendRow(typename TypeTraits<ArrowType_>::BuilderType& builder,       \
+                            CType_ cell) {                                               \
+      return builder.Append(cell);                                                       \
+    }                                                                                    \
+    template <typename Range>                                                            \
```

Comment:
I don't think it's appropriate to make this a member of the ConversionTraits struct unless you standardize on std::vector for your cell range concept; it makes adding new implementations very verbose and detecting whether the functionality is available is now impossible at compile time:
```c++
static_assert(
    HasAppendMultipleRows<bool, BooleanBuilder, MemoryPool*>::value,
    "memory pool shouldn't be usable as a boolean cell range");
```

Please make this a non member method or disallow non-vector cell ranges

Reply:
Yes, that's a good point. How about using an enabler:
`template <typename Range, typename = enable_if<is_same<decay<Range>, std::vector<CType_>>>`

This addresses HasAppendMultipleRows issue. Could you explain a bit more what you mean by adding new implementations?

Reply:
I think I got what you meant. I will try putting this functionality as a free function.